### DISCUSSION TITLE
Getting ObjectDeletedError when accessing record after commit

### DISCUSSION THREAD

**Discussion by jtay-pm** [sqlalchemy/sqlalchemy#10640] [Usage Questions]
```
hh_info_rec = HOUSEHOLD_INFOS(**fields)
sesh.add(hh_info_rec)
sesh.commit()
# creating history tables here using the same session
hh_info_history = HOUSHOLD_INFOS_HISTORY(
    hh_info_id=hh_info_rec.hh_info_id, # throwing an error ObjectDeletedError
    ...
)
```
For some reason, after committing the session, when i try to reference an attribute from the created record instance, sqlalchemy throws this error
```sqlalchemy.orm.exc.ObjectDeletedError: Instance '<household_infos at 0x7f4c30cc08b0>' has been deleted, or its row is otherwise not present.```

This started happening recently and wasn't happening in the past. It also happens on records with a specific tenant id but works for other records with another tenant id. We have row level security enabled on this tenant id column for this table as well and we are sure that the session is executing with the correct permissions and correct tenant id.

Printed the postgresql DB logs and postgresql wasn't throwing any errors when trying to read `hh_info_rec.hh_info_id` but the above transaction gets rolled back because of this error thrown by sqlalchemy. Any ideas here?

Flask-SQLAlchemy==2.5.1
SQLAlchemy==1.4.46
python: 3.8

**Comment by jtay-pm** [comment]
From the postgresql logs, i noticed a pattern..
for failures:
[5328]:LOG:  statement: BEGIN
[5328]:LOG:  statement: SET app.tenant_id='7abf...';
[5328]: INSERT INTO household_infos
[5328]:LOG:  statement: COMMIT
[5332]:LOG:  statement: BEGIN
[5332]:LOG:  statement: SELECT household_infos.hh_info_id ... 
[5332]:LOG:  statement: ROLLBACK

for success:
[8054]:LOG:  statement: BEGIN
[8054]:LOG:  statement: SET app.tenant_id='7ab...';
[8054]:LOG:  statement: INSERT INTO household_infos
[8054]:LOG:  statement: COMMIT
[5328]:LOG:  statement: BEGIN
[5328]:LOG:  statement: SELECT household_infos.hh_info_id ...
[5328]:LOG:  statement: INSERT INTO household_infos_history // no rollback occurs

Is it problematic that the session instance `5328` is somehow reused for another transaction?

**Reply by CaselIT** [reply]
Sqlalchemy uses connection pooling, so the same connection can be reused multiple times.

Also once the session is committed the connection gets restored to the pool, so new operations will use a new connection from the pool.

If tou remote the commit and use a flush() instead, does it work?

**Reply by jtay-pm** [reply]
That works since the same session is persisted throughout. But we have a use case for commit because we have a function that creates a db records, then after that updates the db record. In that case, we need to use sesh.commit() after creating the db records so that the update can be executed.

Problem is after using `sesh.commit()`, when sqlalchemy uses a new connection from the pool, our set up to `[8054]:LOG: statement: SET app.tenant_id='7ab...';` in the previous session is not carried over to this new session.

But from the code, reading this make you think that you are using the same session throughout since `sesh` is passed through so it would be expected that our setting of app.tenant_id on our original session holds. Is there anyway to do this? Or do we have to make sure that whenever we use `sesh.commit()` we have to manually reset the app.tenant_id again?
```
with session_scope() as sesh:
    create_db_records(sesh)
    sesh.commit()
    ... do something else...
    update_db_records(sesh)
```

**Reply by jtay-pm** [reply]
i saw that we can do set `expire_on_commit=False` so that our session doesn't expire. What's the side effect or possible consequences of doing so?

https://docs.sqlalchemy.org/en/20/orm/session_basics.html#committing
> Finally, all objects within the [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) are [expired](https://docs.sqlalchemy.org/en/20/glossary.html#term-expired) as the transaction is closed out. This is so that when the instances are next accessed, either through attribute access or by them being present in the result of a SELECT, they receive the most recent state. This behavior may be controlled by the [Session.expire_on_commit](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.expire_on_commit) flag, which may be set to False when this behavior is undesirable.

**Reply by CaselIT** [reply]
> i saw that we can do set expire_on_commit=False so that our session doesn't expire

that's does not influence the behaviour of the connection

> But from the code, reading this make you think that you are using the same session throughout since sesh is passed through so it would be expected that our setting of app.tenant_id on our original session holds. Is there anyway to do this? Or do we have to make sure that whenever we use sesh.commit() we have to manually reset the app.tenant_id again?

the session stays the same, but a session can use multiple connection during its life.

I'm not sure how you are setting your tenant id, but there are events that allow you to set and reset it when the connection start a transaction etc

alternatively you can create a connection manually, set the identity and pass it to the session using `with Session(connection) as session:`

**Comment by zzzeek** [comment]
I would debug this from the client perspective of the SQL.  set [echo='debug'](https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging) and then reproduce the issue.  When it says "objectdeletederror", the logs will show the SELECT statement being emitted and the rows that were recieved back.  you can look up above to see whatever interaction preceded that.

**Comment by zzzeek** [comment]
> That works since the same session is persisted throughout. But we have a use case for commit because we have a function that creates a db records, then after that updates the db record. In that case, we need to use sesh.commit() after creating the db records so that the update can be executed.
> 
> Problem is after using `sesh.commit()`, when sqlalchemy uses a new connection from the pool, our set up to `[8054]:LOG: statement: SET app.tenant_id='7ab...';` in the previous session is not carried over to this new session.
> 
> But from the code, reading this make you think that you are using the same session throughout since `sesh` is passed through so it would be expected that our setting of app.tenant_id on our original session holds. Is there anyway to do this? Or do we have to make sure that whenever we use `sesh.commit()` we have to manually reset the app.tenant_id again?


you need to use [events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#session-transaction-events) for this kind of thing, if you want a connection-level setting to persist through a certain scope, and then be cleared afterwards.

This is how you do this, setting events on an instance of `Session`:

```py

def set_row_level_thing(session, some_identifier):
    @event.listen(session, "after_begin")
    def _set_rowlevel_security(session, transaction, connection):
        connection.exec_driver_sql(< do PG row level security query here >)

    @event.listen(session, "after_rollback")
    @event.listen(session, "after_commit")
    def _reset_rowlevel_security(session, transaction, connection):
        connection.exec_driver_sql(< un-set-row lkevel security query here >)
```

another way would be to use just "after_begin" for the session, and to use a connection pool level event to reset row level security things.  Below I am setting events for the class of `Session`, using `.info` to track desired row level settings:

```py

from sqlalchemy.orm import Session

engine = create_engine("...")


@event.listen(Session, "after_begin")
def _set_rowlevel_security(session, transaction, connection):
    row_level_token = session.info.get("some_row_level_token", None)
    if row_level_token:
        connection.exec_driver_sql(< do PG row level security query here >)

@event.listen(engine, "checkin")
def _reset_all_rowlevel_security(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute(< reset all PG row level security here >)
    cursor.close()

```

above you would set up the row level token like this:

```py
with Session(e) as sess:
    sess.info["some_row_level_token"] = "my_token"
    
    # ... use session
```


still another way to do this.   Bind the `Session` to a connection directly, then you get the same connection every time:


```py

with engine.connect() as conn:
    conn.execute(<set PG row level security thing>)
    try:
        with Session(conn) as sess:
           # work with Session
    finally:
        conn.execute(<reset PG row level security thing>)
```

All the above patterns can be made succinct using custom context managers and/or function decorators.

**Reply by jtay-pm** [reply]
Thank you!! These suggestions look promising. We actually went to monkeypatch `session.commit()` instead to set the app.tenant_id after a commit. But using these event listeners seems like a better approach.

> Bind the Session to a connection directly, then you get the same connection every time

Are there any side effects or consequences with using the same connection for the same session everytime? Will there be performance implications doing so?

**Reply by CaselIT** [reply]
> But using these event listeners seems like a better approach.

definetly, since rollback or close will also cause the same behaviour

> Are there any side effects or consequences with using the same connection for the same session everytime? Will there be performance implications doing so?

no, you just have to be sure properly manage both the session and the connection (meaning don't leave the connection open indefinitely)

**Reply by dnationalwala-pm** [reply]
```
@event.listen(engine, "checkin")
def _reset_all_rowlevel_security(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute(< reset all PG row level security here >)
    cursor.close()
```

We are sometime getting the error: `AttributeError: 'NoneType' object has no attribute 'cursor'`
is there a way we can check if dbapi_connection is valid before accessing cursor?

**Reply by CaselIT** [reply]
Can you provide the full stack trace?

**Reply by zzzeek** [reply]
> is there a way we can check if dbapi_connection is valid before accessing cursor?

yes, check that it's not None:

```py
    if dbapi_connection is None:
         return
```

**Reply by jtay-pm** [reply]
What exactly does it mean when `dbapi_connection` is `None` btw? We added logic to make sure when the connection is `checkin`, we reset the row level security settings so that the reused connection will be clean. If the connection we received is `None` what does that mean? Just wanted to make sure that the connection is not released back to the connection pool without being resetted with our `reset_rowlevel_security` logic.

**Reply by zzzeek** [reply]
the docs are at https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.checkin

> Note that the connection may be closed, and may be None if the connection has been invalidated. 

We have a typing error here in that the type signature does not show the DBAPIConnection as optional

**Reply by CaselIT** [reply]
@jtay-pm to reply to your question, when the `dbapi_connection is None` it means that a new connect will be done on checkout, getting a new connection from the driver, so that should be in a clean state without any row security setup

**Reply by jtay-pm** [reply]
Thank you! Checking for None helped to resolve the Attribute error we were initially getting.

However, now we've been seeing some errors on the "after_begin" event. When we tried to use `connection.execute` in the after_begin event, we sometimes get this error `(psycopg2.OperationalError) SSL connection has been closed unexpectedly`. This fails fatally and would trigger an internal server error on our end. 

From our sentry logs, it looks like the connection object provided by the event is closed when we tried to execute set tenant id on it.. how should we handle this gracefully? The connection object exists but the underlying cursor object is closed...
```
@event.listen(Session, "after_begin")
def _set_rowlevel_security(session, transaction, connection):
   # Get tenant_id from session
    tenant_id = session.info.get("tenant_id", None)
    LOGGER.debug('SESSION CONNECTION BEGIN!!! %s', tenant_id)
    if tenant_id:
      # Set tenant id on connection
      connection.execute(text("SET app.tenant_id=:tenant_id"), { "tenant_id": tenant_id })
      ...
```

**Reply by jtay-pm** [reply]
By any chance, would using `pre_ping` in `create_engine` resolve this issue here? https://docs.sqlalchemy.org/en/14/core/pooling.html#pool-disconnects

**Reply by CaselIT** [reply]
it would, but if that's a timeout somewhere you may try to use https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.pool_recycle first to see if you still get that error

**Reply by jtay-pm** [reply]
@CaselIT Adding `pre_ping=True` causes this error: `sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) set_session cannot be used inside a transaction` to trigger all the time whenever i use a `commit()` or `connection.execute()` within a transaction.

I tried looking for previous github issues on this error and saw this (https://github.com/sqlalchemy/sqlalchemy/discussions/8009)  but we are not using concurrent threads so i don't think the solution for setting the thread count to 1 applies to us here..

![image](https://github.com/sqlalchemy/sqlalchemy/assets/77286898/e0fa7dd1-6a6e-4006-84fe-90781098d16e)

**Reply by jtay-pm** [reply]
We are using sqlalchemy 1.4.46 and psycopg2 2.9.9 (latest)

**Reply by dnationalwala-pm** [reply]
@CaselIT We are seeing the same error 
```
set_session cannot be used inside a transaction
```
We upgraded the sqlalchemy version to `2.0.38` and psycopg2 to `2.9.10`
